File: android/nativepreview/src/main/java/com/travelplaner/nativepreview/map/GoogleMapSurface.kt
package com.travelplaner.nativepreview.map
import android.content.Context
import androidx.compose.runtime.*
import androidx.compose.ui.Modifier
import androidx.compose.ui.viewinterop.AndroidView
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.LifecycleEventObserver
import androidx.lifecycle.compose.LocalLifecycleOwner
import androidx.compose.ui.platform.LocalContext
import com.google.android.gms.maps.*
import com.google.android.gms.maps.model.*
import com.travelplaner.nativepreview.domain.*

@Composable
fun GoogleMapSurface(projection:MapProjection,command:CameraCommand?,active:Boolean,onSelect:(MapMarker)->Unit,onReady:()->Unit,modifier:Modifier=Modifier,initialCamera:MapCameraSnapshot?=null,onCameraIdle:(MapCameraSnapshot)->Unit={}) {
    val context=LocalContext.current
    val owner=LocalLifecycleOwner.current
    val currentActive by rememberUpdatedState(active)
    val controller=remember(context) { MapSurfaceController(context,initialCamera) }
    DisposableEffect(controller,owner) {
        controller.view.onCreate(null)
        controller.view.getMapAsync { map -> controller.attach(map) }
        val observer=LifecycleEventObserver { _,_ -> controller.lifecycle(owner.lifecycle.currentState,currentActive) }
        owner.lifecycle.addObserver(observer); controller.lifecycle(owner.lifecycle.currentState,active)
        onDispose { owner.lifecycle.removeObserver(observer); controller.destroy() }
    }
    SideEffect { controller.lifecycle(owner.lifecycle.currentState,active) }
    AndroidView(factory={controller.view},modifier=modifier,update={
        controller.update(projection,command,active,onSelect,onReady,onCameraIdle)
    })
}
private class MapSurfaceController(context:Context,private val initialCamera:MapCameraSnapshot?) {
    val view=MapView(context,GoogleMapOptions().mapToolbarEnabled(false).zoomControlsEnabled(false).compassEnabled(true))
    private var map:GoogleMap?=null
    private val markers=mutableMapOf<String,Marker>()
    private val values=mutableMapOf<String,MapMarker>()
    private var routes=emptyList<List<Coordinate>>()
    private var lines=emptyList<Polyline>()
    private val queue=CameraQueue()
    private var projection=MapProjection(emptyList(),0,emptyList())
    private var active=true; private var started=false; private var resumed=false; private var destroyed=false
    private var select:(MapMarker)->Unit={}; private var ready:()->Unit={}
    private var cameraIdle:(MapCameraSnapshot)->Unit={}
    init { view.addOnLayoutChangeListener { _,_,_,_,_,_,_,_,_ -> consume() } }
    fun attach(map:GoogleMap) {
        if(destroyed) return
        this.map=map
        map.moveCamera(CameraUpdateFactory.newLatLngZoom(LatLng(35.0,135.0),3f))
        initialCamera?.values?.let {v -> map.moveCamera(CameraUpdateFactory.newCameraPosition(CameraPosition(LatLng(v[0],v[1]),v[2].toFloat(),v[4].toFloat(),v[3].toFloat()))) }
        map.setOnCameraIdleListener {
            val p=map.cameraPosition
            MapCameraSnapshot.from(listOf(p.target.latitude,p.target.longitude,p.zoom.toDouble(),p.bearing.toDouble(),p.tilt.toDouble()))?.let(cameraIdle)
        }
        map.setOnMapLoadedCallback { ready(); consume() }
        map.setOnMarkerClickListener { marker -> (marker.tag as? String)?.let(values::get)?.let(select); true }
        map.setOnCameraMoveStartedListener { reason -> if(reason==GoogleMap.OnCameraMoveStartedListener.REASON_GESTURE) queue.cancelPending() }
        reconcile(); consume()
    }
    fun update(projection:MapProjection,command:CameraCommand?,active:Boolean,onSelect:(MapMarker)->Unit,onReady:()->Unit,onCameraIdle:(MapCameraSnapshot)->Unit) {
        this.projection=projection; this.active=active; select=onSelect; ready=onReady
        cameraIdle=onCameraIdle
        if(!active) queue.cancelPending() else command?.let(queue::enqueue)
        reconcile(); consume()
    }
    private fun reconcile() {
        val map=map ?: return
        val desired=projection.markers.associateBy { it.key }
        markers.keys.filter { it !in desired }.forEach { markers.remove(it)?.remove(); values.remove(it) }
        projection.markers.forEach { value ->
            if(values[value.key]!=value) {
                val point=LatLng(value.coordinate.latitude,value.coordinate.longitude)
                val marker=markers[value.key] ?: map.addMarker(MarkerOptions().position(point)) ?: return@forEach
                marker.position=point; marker.title=value.label; marker.tag=value.key
                marker.setIcon(BitmapDescriptorFactory.defaultMarker(if(value.kind==MapMarker.Kind.Saved) BitmapDescriptorFactory.HUE_ROSE else BitmapDescriptorFactory.HUE_CYAN))
                markers[value.key]=marker; values[value.key]=value
            }
        }
        val next=projection.routes.map { route -> route.markers.map { it.coordinate } }
        if(next!=routes) {
            lines.forEach { it.remove() }
            lines=next.map { points -> map.addPolyline(PolylineOptions().addAll(points.map { LatLng(it.latitude,it.longitude) }).color(android.graphics.Color.rgb(65,108,98)).width(5f)) }
            routes=next
        }
    }
    private fun consume() {
        val map=map ?: return
        queue.consume(active && view.width>0 && view.height>0) { command ->
            when(val target=command.target) {
                is CameraTarget.Center -> map.moveCamera(CameraUpdateFactory.newLatLngZoom(LatLng(target.coordinate.latitude,target.coordinate.longitude),target.zoom))
                is CameraTarget.Fit -> MapBounds.fit(target.coordinates)?.let {
                    val bounds=LatLngBounds(LatLng(it.south,it.west),LatLng(it.north,it.east))
                    val padding=(32*view.resources.displayMetrics.density).toInt().coerceAtMost(minOf(view.width,view.height)/4)
                    map.moveCamera(CameraUpdateFactory.newLatLngBounds(bounds,padding))
                }
            }
        }
    }
    fun lifecycle(state:Lifecycle.State,active:Boolean) {
        if(destroyed) return
        val shouldStart=state.isAtLeast(Lifecycle.State.STARTED)
        val shouldResume=state.isAtLeast(Lifecycle.State.RESUMED) && active
        if(resumed && !shouldResume) { view.onPause(); resumed=false }
        if(started && !shouldStart) { view.onStop(); started=false }
        if(!started && shouldStart) { view.onStart(); started=true }
        if(!resumed && shouldResume) { view.onResume(); resumed=true }
    }
    fun destroy() {
        if(destroyed) return
        lifecycle(Lifecycle.State.CREATED,false); destroyed=true
        map?.setOnMarkerClickListener(null); map?.setOnMapLoadedCallback(null); map?.setOnCameraMoveStartedListener(null)
        map?.setOnCameraIdleListener(null)
        markers.values.forEach { it.remove() }; lines.forEach { it.remove() }; markers.clear(); values.clear()
        view.onDestroy(); map=null
    }
}

File: ios/NativePreview/Map/GoogleMapSurface.swift
import SwiftUI
import GoogleMaps
import NativeCore

struct GoogleMapSurface: UIViewRepresentable {
    let projection:MapProjection
    let command:CameraCommand?
    let active:Bool
    let onSelect:(MapMarker)->Void
    let onReady:()->Void
    var initialCamera:MapCameraSnapshot?=nil
    var onCameraIdle:(MapCameraSnapshot)->Void={_ in}
    func makeCoordinator() -> Coordinator { Coordinator(self) }
    func makeUIView(context:Context) -> GMSMapView {
        // Parent only mounts after MapConfiguration.configure has accepted a nonblank key.
        let options=GMSMapViewOptions()
        options.camera=GMSCameraPosition(latitude:35,longitude:135,zoom:3)
        if let v=initialCamera?.values {options.camera=GMSCameraPosition(latitude:v[0],longitude:v[1],zoom:Float(v[2]),bearing:v[3],viewingAngle:v[4])}
        let map=GMSMapView(options:options)
        map.delegate=context.coordinator; map.settings.myLocationButton=false; map.settings.compassButton=true
        context.coordinator.update(self,map:map); return map
    }
    func updateUIView(_ map:GMSMapView,context:Context) { context.coordinator.update(self,map:map) }
    static func dismantleUIView(_ view:GMSMapView,coordinator:Coordinator) {
        view.delegate=nil
        coordinator.markers.values.forEach { $0.map=nil }; coordinator.lines.forEach { $0.map=nil }
        coordinator.markers=[:]; coordinator.lines=[]
    }
    final class Coordinator:NSObject,GMSMapViewDelegate {
        var owner:GoogleMapSurface
        var markers=[String:GMSMarker](), snapshots=[String:MapMarker]()
        var lines=[GMSPolyline](), routeCoordinates=[[Coordinate]]()
        var queue=CameraQueue(), ready=false
        init(_ owner:GoogleMapSurface) { self.owner=owner }
        func update(_ owner:GoogleMapSurface,map:GMSMapView) {
            self.owner=owner
            if !owner.active { queue.cancelPending() }
            let desired=Dictionary(uniqueKeysWithValues:owner.projection.markers.map { ($0.key,$0) })
            for key in markers.keys.filter({desired[$0]==nil}) { markers.removeValue(forKey:key)?.map=nil; snapshots.removeValue(forKey:key) }
            for value in owner.projection.markers where snapshots[value.key] != value {
                let marker=markers[value.key] ?? GMSMarker()
                marker.position=CLLocationCoordinate2D(latitude:value.coordinate.latitude,longitude:value.coordinate.longitude)
                marker.title=value.label; marker.userData=value.key
                marker.icon=GMSMarker.markerImage(with:value.kind == .saved ? .systemPink : .systemTeal)
                marker.map=map; markers[value.key]=marker; snapshots[value.key]=value
            }
            let routes=owner.projection.routes.map { $0.markers.map(\.coordinate) }
            if routes != routeCoordinates {
                lines.forEach { $0.map=nil }
                lines=routes.map { coordinates in
                    let path=GMSMutablePath(); coordinates.forEach { path.addLatitude($0.latitude,longitude:$0.longitude) }
                    let line=GMSPolyline(path:path); line.strokeColor = .systemTeal; line.strokeWidth=3; line.map=map; return line
                }
                routeCoordinates=routes
            }
            if let command=owner.command { queue.enqueue(command) }
            consume(map)
        }
        private func consume(_ map:GMSMapView) {
            queue.consume(ready:ready && owner.active && map.bounds.width > 0 && map.bounds.height > 0) { command in
                switch command.target {
                case .center(let point,let zoom):
                    map.moveCamera(.setTarget(CLLocationCoordinate2D(latitude:point.latitude,longitude:point.longitude),zoom:Float(zoom)))
                case .fit(let points):
                    if let bound=MapBounds.fit(points) {
                        let sw=CLLocationCoordinate2D(latitude:bound.south,longitude:bound.west), ne=CLLocationCoordinate2D(latitude:bound.north,longitude:bound.east)
                        map.moveCamera(.fit(GMSCoordinateBounds(coordinate:sw,coordinate:ne),withPadding:min(40,min(map.bounds.width,map.bounds.height)/4)))
                    }
                }
            }
        }
        func mapViewDidFinishTileRendering(_ mapView:GMSMapView) { ready=true; owner.onReady(); consume(mapView) }
        func mapView(_ mapView:GMSMapView,idleAt position:GMSCameraPosition) {
            if let snapshot=MapCameraSnapshot(values:[position.target.latitude,position.target.longitude,Double(position.zoom),position.bearing,position.viewingAngle]) {owner.onCameraIdle(snapshot)}
        }
        func mapView(_ mapView:GMSMapView,willMove gesture:Bool) { if gesture { queue.cancelPending() } }
        func mapView(_ mapView:GMSMapView,didTap marker:GMSMarker) -> Bool {
            guard let key=marker.userData as? String,let value=snapshots[key] else { return false }; owner.onSelect(value); return true
        }
    }
}
